crates/sui-faucet/src/server.rs

```diff
@@ -35,12 +35,203 @@ use tracing::{info, warn};
 use uuid::Uuid;
 
 use crate::faucet::Faucet;
+use serde::Deserialize;
+use std::sync::Mutex;
+
+use anyhow::bail;
+use once_cell::sync::Lazy;
+
+/// Interval to cleanup expired tokens
+const CLEANUP_INTERVAL: u64 = 60; // 60 seconds
+/// Maximum number of requests per IP address
+const MAX_REQUESTS_PER_IP: u32 = 3;
+/// Interval to reset the request count for each IP address
+const RESET_TIME_INTERVAL_SECS: u64 = 12 * 3600; // 12 hours
+const CLOUDFLARE_URL: &str = "https://challenges.cloudflare.com/turnstile/v0/siteverify";
+const FAUCET_WEB_APP_URL: &str = "https://faucet.sui.io"; // make this lazy static env?
+
+static DISCORD_BOT: Lazy<Option<String>> = Lazy::new(|| std::env::var("DISCORD_BOT").ok());
+
+static TURNSTILE_SECRET_KEY: Lazy<Option<String>> =
+    Lazy::new(|| std::env::var("TURNSTILE_SECRET_KEY").ok());
+
+type IPAddr = String;
+
+/// Keep track of every IP address' requests.
+#[derive(Debug)]
+struct RequestsManager {
+    pub data: Mutex<BTreeMap<IPAddr, RequestInfo>>,
+    reset_time_interval_secs: u64,
+    max_requests_per_ip: u32,
+}
+
+/// Request's metadata
+#[derive(Debug, Clone)]
+struct RequestInfo {
+    expires_at: u64,
+    requests_used: u32,
+    total_requests_allowed: u32,
+}
+
+/// Struct to deserialize token verification response from Cloudflare
+#[derive(Deserialize, Debug)]
+struct TokenVerification {
+    success: bool,
+    #[serde(rename = "error-codes")]
+    error_codes: Vec<String>,
+}
+
+impl RequestsManager {
+    /// Initialize a new RequestsManager the default values.
+    fn new() -> Self {
+        Self {
+            data: Mutex::new(BTreeMap::new()),
+            reset_time_interval_secs: RESET_TIME_INTERVAL_SECS,
+            max_requests_per_ip: MAX_REQUESTS_PER_IP,
+        }
+    }
+
+    #[cfg(test)]
+    fn new_with_limits(max_requests_per_ip: u32, reset_time_interval_secs: u64) -> Self {
+        Self {
+            data: Mutex::new(BTreeMap::new()),
+            reset_time_interval_secs,
+            max_requests_per_ip,
+        }
+    }
+
+    /// Get the current timestamp in seconds
+    fn current_timestamp_secs() -> u64 {
+        SystemTime::now()
```

Comment:
System time is usually the wrong thing to use to measure timeouts because it's not monotonic -- the system clock could go back in time for any number of reasons and you don't want that to affect the timeouts you've set-up -- consider using `std::time::Instant` instead (`Instant::now()` is monotonic).

Also, is it possible to avoid losing the type safety by unwrapping the type into a `u64`? The ideal way to implement something like this is to remember the timestamp as an `Instant` and then you can call `Instant::now().duration_since(...)` to get the gap between the two values as a `Duration` which you can compare with your expiry.